modules/clightning/clightning_lib.cpp

```diff
@@ -0,0 +1,81 @@
+#define template c_template  // avoid C++ keyword conflict just during includes
+
+extern "C" {
+    #include "common/bolt11.h"
+    #include "bitcoin/pubkey.h"
+    #include "common/node_id.h"
+    #include "common/utils.h"
+    #include <bitcoin/chainparams.h>
+    #include <ccan/tal/tal.h>
+}
+
+#undef template
+
+#include "clightning_lib.h"
+#include <string>
+#include <sstream>
+#include <iomanip>
+#include <cstring>
+#include <vector>
+#include <memory>
+#include <iostream>
+
+struct TalFree {
+    void operator()(void* ptr) const { tal_free(ptr); }
+};
+
+std::string hex_encode(const unsigned char* data, size_t len) {
+    std::ostringstream oss;
+    oss << std::hex << std::setfill('0');
+    for (size_t i = 0; i < len; ++i) {
+        oss << std::setw(2) << static_cast<int>(data[i]);
+    }
+    return oss.str();
+}
+
+std::string clightning_des_invoice(const std::string& input) {
+    char* fail = nullptr;
+    const struct chainparams* params = chainparams_for_network("bitcoin");
+
+    std::unique_ptr<bolt11, TalFree> invoice(
+        bolt11_decode(nullptr, input.c_str(), nullptr, nullptr, params, &fail)
+    );
+
+    if (!invoice) {
+        if (fail) tal_free(fail);
```

Comment:
It's OK to do `tal_free(NULL)`.

```suggestion
        tal_free(fail);
```

Reply:
Done